### ISSUE TITLE
Why does the neural_network.py use log softmax rather than softmax?

### ISSUE THREAD

**Issue by MrDavidG** [mravanelli/pytorch-kaldi#117]
Hi! I am reading the source code and I'm little confused why  ```neural_network.py``` use ```LogSoftmax``` rather than ```Softmax```?
The code in ```neural_networks.py``` is 
```
def act_fun(act_type):
    if act_type == "relu":
        return nn.ReLU()

    if act_type == "tanh":
        return nn.Tanh()

    if act_type == "sigmoid":
        return nn.Sigmoid()

    if act_type == "leaky_relu":
        return nn.LeakyReLU(0.2)

    if act_type == "elu":
        return nn.ELU()

    if act_type == "softmax":
        return nn.LogSoftmax(dim=1)

    if act_type == "linear":
        return nn.LeakyReLU(1)  # initializzed like this, but not used in forward!
```

In the example dense net, pytorch-kaldi uses the output (log softmax) of net and the negative log-likelihood loss(NLL). That means it would use the log softmax in decoding and all output would be negative, which is contrary to what I thought before: **use ```softmax``` function to obtain a probability-like output**.  

Besides, I have also check the file ```/home/qiaoyangliu/Program/pytorch-kaldi/exp/TIMIT_MLP_basic/exp_files/forward_TIMIT_test_ep22_ck0_out_dnn1_to_decode.ark``` and all values are negative. So here is my problem:

 - why does the network use log softmax as the final activation function? Is it a requirement of kaldi command? Because it seems like the cmd ```latgen-faster-mapped``` in ```decode_dnn.sh``` uses this output (all values are negative) to decode.

**Comment by xingchensong** [comment]
I agree with u , it's because the cmd  `latgen-faster-mapped` need log-likelihood.

**Comment by xingchensong** [comment]
![image](https://user-images.githubusercontent.com/13466943/63839761-753ffc00-c9b2-11e9-9ca5-c40df5c174a6.png)
